joshua/joshua_model.py

```diff
@@ -736,8 +736,9 @@ def _insert_results(tr,
 
     if max_runs > 0:
         # This is a snapshot read so that two insertions don't conflict.
+        started = _get_snap_counter(tr, ensemble_id, 'started')
         ended = _get_snap_counter(tr, ensemble_id, 'ended')
-        if ended >= max_runs:
+        if ended >= max(max_runs, started):
```

Comment:
I don't understand the code base well enough to know whether this will work or not, but what happens if an agent dies after incrementing `started` to a value larger than `max_runs`?

An alternative to this fix could also be to make the start transaction a bit more complex and increment `max_runs` whenever we overshoot?

Reply:
Does a running test exist in the data model under any unique key?  Is there a lease the agent renews to indicate that the test it is running is still in flight?  Or is it just counters and a lot of hope?

Reply:
It seems like it would not be too much of a bottleneck to serialize test starts under a job on committing an FDB transaction that checked `started` vs `max_runs` and if not met then increment `started` and `commit()`, with a read conflict on `started`, and only start the test after the commit is successful.

This will fully prevent overruns, but now to address the agent death issue where `ended` and `pass or fail` is never incremented.

I have a suggestion that isn't the greatest but it will _eventually_ get the job into the right state.  If we track `lastStartVersion` for the job ID in the transaction above, then agents can periodically check the database for job IDs which have
```
started >= max_runs && tr.getReadVersion() > (lastStartVersion + (1000000 * timeout * 1.1))
// in this model started should never *exceed* max_runs of course...
```
This will keep the job 'alive' to collect all results including timeouts of the started tests until so much time has passed that any timeouts should definitely have occurred by now.

Reply:
Sorry, I already made a significant change to this PR.
It shouldn't overshoot now.
It's slightly different mechanism from what Steve suggested, but basically, it won't start jobs more than max_runs.
When it detects "cancelled" job = dead agent, it simply ignores that run even if it comes back later.
This is not a test failure. It's just the agent hiccup, so it's safe to exclude.

Reply:
How is a dead agent detected?  Do agents check in periodically?

Reply:
Each agent will update the heartbeat every second. When `started` >= `max_runs` (really, ==), idle agents check for the WIP tests and see if it's getting heartbeats. When it finds it has not received heartbeats for more than 10 seconds, it will steal the job.

Reply:
When stealing, it also decrements the `started` counter, so Joshua completely ignores the "dead" job (even if it comes back later, it's already taken over by another agant)

Reply:
Nice!  This sounds essentially equivalent and failure-proof to what I was thinking.

Reply:
(by failure proof I mean we won't have jobs stuck in a zombie state forever with no launchings, nothing running)

Reply:
Regarding the 10s heartbeat, it might be a good idea increase that to 30 just because if the cluster ever hiccups badly we would be most likely to lose the longest running tests, which are exactly the ones we want to stop missing.